From 8f2ca31010c63958522af09aba15014623bd52bc Mon Sep 17 00:00:00 2001
From: Lauri Myllari <lauri.myllari@gmail.com>
Date: Mon, 27 Jul 2015 22:02:32 -0700
Subject: [PATCH] lut: fix 3dlut texture coordinates

---
 system/shaders/output.glsl | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/system/shaders/output.glsl b/system/shaders/output.glsl
index 31013c4..e629231 100644
--- a/system/shaders/output.glsl
+++ b/system/shaders/output.glsl
@@ -12,7 +12,8 @@ void main()
   vec4 rgb        = process();
 
 #if (XBMC_3DLUT)
-  rgb             = texture3D(m_CLUT, rgb.rgb);
+  // FIXME: can this be optimized?
+  rgb             = texture3D(m_CLUT, (rgb.rgb*255.0 + (0.5, 0.5, 0.5)) / 256.0);
 #endif
 
 #if (XBMC_FULLRANGE)
-- 
2.4.3

